onnxruntime/core/framework/sequential_executor.cc

```diff
@@ -122,11 +123,218 @@ static Status ReleaseNodeMLValues(ExecutionFrame& frame,
                                   const SequentialExecutionPlan::NodeExecutionPlan& node_exec_plan,
                                   const logging::Logger& logger);
 
+static Status FetchOutputsFromExecutionFrame(ExecutionFrame& frame, bool release_outputs,
+                                             std::vector<OrtValue>& fetches, const logging::Logger& logger) {
+  VLOGS(logger, 1) << "Fetching output.";
+  ORT_RETURN_IF_ERROR(frame.GetOutputs(fetches, release_outputs));
+  VLOGS(logger, 1) << "Done with execution.";
+  return Status::OK();
+}
+
+static Status SynchronizeNodeInputs(const SequentialExecutionPlan& seq_exec_plan, size_t program_counter,
+                                    const SessionState& session_state, ExecutionFrame& frame, bool terminate_flag,
+                                    const logging::Logger& logger) {
+  const auto& node_exec_plan = seq_exec_plan.execution_plan[program_counter];
+  auto node_index = node_exec_plan.node_index;
+  const auto& graph_viewer = session_state.GetGraphViewer();
+  const auto& node = *graph_viewer.GetNode(node_exec_plan.node_index);
+  auto p_op_kernel = session_state.GetKernel(node_index);
+  if (p_op_kernel == nullptr)
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Got nullptr from GetKernel for node: ",
+                           node.Name());
+
+  OpKernelContextInternal op_kernel_context(session_state, frame, *p_op_kernel, logger, terminate_flag);
+  int queue_id = p_op_kernel->KernelDef().ExecQueueId();
```

Comment:
This (unrelated?) refactoring might cause a performance regression, because it creates 5 `OpKernelContextInternal` objects for each node executed, versus (I think) 1 before.

Revert?

Reply:
I have been thinking about this too and can further refactor to prevent redundant work. 

---
In reply to: [600932445](https://github.com/microsoft/onnxruntime/pull/7077#discussion_r600932445) [](ancestors = 600932445)